tests/test_tools.py

```diff
@@ -0,0 +1,117 @@
+import pytest
+import logging
+from pkg_resources import resource_filename
+from wps_tools.utils import (
+    is_opendap_url,
+    get_filepaths,
+    collect_output_files,
+    build_meta_link,
+    log_handler,
+)
+from wps_tools.testing import (
+    local_path,
+    opendap_path,
+    run_wps_process,
+)
+from processes.wps_say_hello import SayHello
+
+nc_file = "gdd_annual_CanESM2_rcp85_r1i1p1_1951-2100.nc"
+
+
+class NCInput:  # For testing 'get_filepaths'
+    def __init__(self, url="", file=""):
+        self.url = url
+        self.file = file
+
+
+class Response:  # For testing 'log_handler'
+    def __init__(self):
+        self.message = ""
+        self.status_percentage = 0
+
+    def update_status(self, message, status_percentage):
+        self.message = message
+        self.status_percentage = status_percentage
+
+
+# Test 'testing' functions
+def test_local_path():
+    assert f"tests/data/{nc_file}" in local_path(nc_file)
+
+
+def test_opendap_path():
+    assert f"datasets/TestData/{nc_file}" in opendap_path(nc_file)
+
+
+def test_run_wps_process():
+    params = "name=PCIC"
+    run_wps_process(SayHello(), params)
+
+
+# Test 'utils' functions
+@pytest.mark.online
+@pytest.mark.parametrize(
+    ("url"), [opendap_path(nc_file), local_path(nc_file),],
+)
+def test_is_opendap_url(url):
+    if "docker" in url:
+        assert is_opendap_url(url)
+    else:
+        assert not is_opendap_url(url)
+
+
+@pytest.mark.online
+@pytest.mark.parametrize(
+    ("nc_input"),
+    [
+        [NCInput(file=local_path(nc_file))],
+        [NCInput(url=opendap_path(nc_file))],
+        [NCInput(file=local_path(nc_file)), NCInput(url=opendap_path(nc_file))],
+    ],
+)
+def test_get_filepaths(nc_input):
+    for input in nc_input:
+        if input.url != "":
+            assert is_opendap_url(input.url)
+
+    paths = get_filepaths(nc_input)
+    assert len(paths) == len(nc_input)
+    for path in paths:
+        assert nc_file in path
+
+
+@pytest.mark.parametrize(("varname"), ["tiny"])
+@pytest.mark.parametrize(("outdir"), [resource_filename(__name__, "data")])
+def test_collect_output_files(varname, outdir):
+    outfiles = collect_output_files(varname, outdir)
+    assert len(outfiles) == 2
+    assert outfiles == ["tiny_daily_prsn.nc", "tiny_daily_pr.nc"]
```

Comment:
I am getting a failure running this:
```
varname = 'tiny', outdir = '/home/sangwonl/wps-tools/tests/data'

    @pytest.mark.parametrize(("varname"), ["tiny"])
    @pytest.mark.parametrize(("outdir"), [resource_filename(__name__, "data")])
    def test_collect_output_files(varname, outdir):
        outfiles = collect_output_files(varname, outdir)
        assert len(outfiles) == 2
>       assert outfiles == ["tiny_daily_prsn.nc", "tiny_daily_pr.nc"]
E       AssertionError: assert ['tiny_daily_...aily_prsn.nc'] == ['tiny_daily_..._daily_pr.nc']
E         At index 0 diff: 'tiny_daily_pr.nc' != 'tiny_daily_prsn.nc'
E         Full diff:
E         - ['tiny_daily_prsn.nc', 'tiny_daily_pr.nc']
E         + ['tiny_daily_pr.nc', 'tiny_daily_prsn.nc']

tests/test_tools.py:88: AssertionError
```
Although the `outfiles` variable has the value that we desire, python thinks it is different since it is comparing in list data structure, which is index dependant. You might want to convert it to `set` or find comparison method other than `==`

Reply:
Yeah, I'll definitely do that. For some reason, it didn't occur to me that the order could change if someone else tests it.